corehq/apps/userreports/models.py

```diff
@@ -214,6 +234,16 @@ def deleted_filter(self, document):
         filter_fn = self._get_deleted_filter()
         return filter_fn and filter_fn(document, EvaluationContext(document, 0))
 
+    @property
+    def has_validations(self):
+        return len(self.validations) > 0
+
+    def validate_document(self, document):
+        for validation in self.validations:
+            validate_fn = FilterFactory.from_spec(validation.expression, context=self.get_factory_context())
```

Comment:
I suspect you might want to memoize these on the data source. these factory functions can get expensive

Reply:
thanks for the tip https://github.com/dimagi/commcare-hq/pull/23799/commits/e69ed227cfd6d7c9c374e1e3bcb00c4473d2a0ae this would have been really expensive to do every document